docs/architecture-decisions/2024-06-06_arc4-types.md

```diff
@@ -0,0 +1,84 @@
+
+## Options
+
+### Option 1 - Implement only ARC4 types
+
+Algorand TypeScript defines all types specified by ARC4 and their in memory representation matches that of the ARC4 encoding
+
+Pros:
+ - Zero cost decoding of ABI args (other than packed args for methods using more than 15 arguments)
+ - When storing a value in global/local/box storage - the binary format is already well-defined by arc4
+
+Cons:
+ - Contracts that frequently mutate dynamic structs will consume a large op code budget
+ - Math operators that work on byte arrays are expensive (10x) so smaller (<=64 bit) uints would need to be converted to uint64 for math, then back again.
+
+### Option 2 - Implement only 'native' types
+
+Algorand TypeScript defines all types specified by ARC4 but chooses its own in memory representation. Values are encoded/decoded automatically at applicable boundaries.
+
+This is the approach taken by TealScript
+
+Pros:
+ - In memory representation can be optimised
+
+Cons:
+ - In some scenarios a value may be unnecessarily decoded and re-encoded when it could have been passed through untouched.
+ - A number of types defined by ARC4 do not make sense to implement natively (eg. UFixedNxM) as the avm has no support for decimal math.
+ 
+### Option 3 - Implement 'native' types which make sense + implement a full set of ARC4 types
+
+Algorand TypeScript defines native types which make sense and excludes ones which don't (eg. UFixedNxM), but also implements all arc4 types in a separate module. It's possible to convert between the two - where such a conversion exists. Native types can have optimised in memory representations whilst arc4 types follow prescribed encoding.
+
+This is the approach taken by Algorand Python.
+
+Pros:
+ - Zero cost decoding of ABI args available as an option. (by opting to use the arc4 type)
+ - "native" types can be used in ABI args and the compiler will automatically decode them
+ - Arc4 types can optionally be used as a serialization format for box/global/local storage
+ - Developer has control to choose the solution which works best for their scenario rather than having one prescribed by the compiler
+ 
+
+Cons:
+ - Several types will exist twice (string/uint64/bool etc) and this could be confusing to new developers who aren't fully across arc4
+ - Developer will need to be aware of arc4 encoding/decoding and the implications of it.
+
+## Preferred option
+
+Option 1 can be excluded as it doesn't allow us to optimise the in memory representation of types independently to that which is prescribed by the arc4 encoding spec.
+
+Option 2 is feasible however puya would need to implement math operators on sized uints < 64bit and that effort would largely be a waste of time as this math would be more expensive than 64bit math. TealScript does not face this problem as it does not attempt to maintain semantic compatability. This option also does not allow power users to deal directly with encoded data. 
```

Comment:
> TealScript does not face this problem as it does not attempt to maintain semantic compatability.

I'm confused by this statement. It uses uint64 for small uints but I don't see how that relates to semantic compatibility

> This option also does not allow power users to deal directly with encoded data.

Sure it does, for example in TEALScript there is a `rawBytes` function which will give you the encoded data.

Reply:
TealScript allows a `uint<N>` expression to exceed the bounds of an n-bit number and only performs bounds checking when it's assigned to a `uint<N>` variable again. (Side note, is this checking new? I thought it previously only did it at serialization boundaries). This is not semantically compatible with the source language where a `number` can't temporarily exceed it's safe limits without ill effects (in the case of floating point math it would be loss of precision rather than an overflow error).

Eg. this should cause an overflow error but it does not 
```ts
    const x: uint8 = 255;
    const y: uint8 = 255;
    const z: uint8 = x + y - x;
```

For your second point, from memory TealScript truncates the length header from dynamic arrays of uints and only restores it at serialization boundaries so `rawBytes(x)` on a dynamic array will not match the raw bytes actually passed in. The fact that it's even close to the _actual_ raw bytes though is merely a coincidence of TealScript's in memory representation being _close_ to the arc4 encoding. This will NOT be the case in Puya meaning there won't be a zero/near zero cost method of accessing the raw bytes of an ABI param unless we implement a separate type to represent it as such. Of course one can always access application args directly to get the raw bytes, but by that time you've already paid the cost of decoding the data into our in-memory representation _unless_ we have a separate type which can be used in the method args which represents the data in it's encoded format.

Reply:
It doesn't matter if uint8 is a native JavaScript or TypeScript type. uint8 in itself has semantic meaning and that meaning should be respected by the compiler (by either error-ing or cycling or clamping overflows). Obviously we can implement this in the compiler, but there is a cost and that cost may not be worth it when we could just guide devs to use uint64 math.